core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1207 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the 'License');
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an 'AS-IS' BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit test for StateGraphLayoutService.
+ */
+
+import { TestBed } from '@angular/core/testing';
+import { AppConstants } from 'app.constants';
+import cloneDeep from 'lodash/cloneDeep';
+
+import { GraphLink, GraphNodes } from 'services/compute-graph.service';
+import { StateGraphLayoutService } from './graph-layout.service';
+
+describe('Graph Layout Service', () => {
+  let sgls: StateGraphLayoutService = null;
+  let nodeData1 = {
```

Comment:
Perhaps give a brief explanation (in comments) of the graph that this complicated dict represents.

Reply:
Added

Reply:
It looks fantastic. Thanks!